### PR TITLE
Added ability to receive return url and store in session

### PR DESCRIPTION
**Pull request by davidmthornton**

### CODE REVIEW

File: app/controllers/JourneySetupController.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        Future.successful(Ok("Received and stored the redirect url: " + extractRedirectUrl(json))).map {
+          result => result.addingToSession(SessionKeys.REDIRECT_URL -> extractRedirectUrl(json))
+        }
+      }
+
+      case _ => Future.successful(BadRequest)
+    }
+  }
+
+  private def extractRedirectUrl(json: JsValue): String = (json \ "returnUrl").as[String]
```
Comment (TomDudley92) [inline app/controllers/JourneySetupController.scala]: Can you add a test for this please.

File: app/controllers/JourneySetupController.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                                       implicit val appConfig: AppConfig)
+  extends AuthorisedController with I18nSupport {
+
+  def journeySetup():Action[AnyContent] = authorisedAction {
+    implicit request =>
+    implicit agent =>
+    request.body.asJson match {
+
+      case Some(json) if extractRedirectUrl(json).nonEmpty => {
+        Future.successful(Ok("Received and stored the redirect url: " + extractRedirectUrl(json))).map {
```
Comment (TomDudley92) [inline app/controllers/JourneySetupController.scala]: This needs testing as it's reducing the overall coverage below the minimum threshold

File: app/controllers/JourneySetupController.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+class JourneySetupController @Inject()(val messagesApi: MessagesApi,
+                                          val authenticate: AuthoriseAsAgentOnly,
+                                          val serviceErrorHandler: ErrorHandler,
+                                          implicit val appConfig: AppConfig) extends FrontendController with I18nSupport {
+
+  def journeySetup():Action[AnyContent] = authenticate.async {
+    implicit request =>
+    request.body.asJson match {
+
+      case Some(json) => {
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you remove redundant braces please

File: app/controllers/JourneySetupController.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            result => {
+              result.addingToSession(SessionKeys.redirectUrl -> url)
+            }
+          }
+        }else{
+          Logger.warn("[JourneySetupController][journeySetup] redirectUrl was empty or an invalid absolute url")
+          Future.successful(BadRequest)
+        }
+      }
+      case None => {
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you remove redundant braces please

File: app/controllers/JourneySetupController.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        Future.successful(BadRequest)
+      }
+    }
+  }
+
+  private[controllers] def extractRedirectUrl(json: JsValue): Option[String] = {
+    try{
+      Some((json \ "redirectUrl").as[String])
+    }catch{
+      case jsEx:JsResultException => {
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you remove redundant braces please

File: test/controllers/JourneySetupControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        status(result) shouldBe Status.OK
+
+      }
+
+      "store the redirect url in session" in {
+
+        mockAgentAuthorised()
+
+        val result = await(TestJourneySetupControllerSpec.journeySetup()(fakeRequestWithJson))
+        val x = await(result map {
```
Comment (HarryGull) [inline test/controllers/JourneySetupControllerSpec.scala]: Can you give this a meaniful name please (e..g sessionUrl)

File: test/controllers/JourneySetupControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+          res => {
+            session(res).get(common.SessionKeys.redirectUrl).get
+          }
+
+        })
+
+        x shouldEqual goodRedirectUrl
```
Comment (HarryGull) [inline test/controllers/JourneySetupControllerSpec.scala]: As above for rename

File: test/controllers/JourneySetupControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (HarryGull) [inline test/controllers/JourneySetupControllerSpec.scala]: Can you add some tests to cover the result being  BAD requests as returned in the code please.

File: app/controllers/JourneySetupController.scala (hunk 3 above)
Comment (davidmthornton) [reply app/controllers/JourneySetupController.scala]: I can't see any redundant braces there?

Comment (davidmthornton) [reply app/controllers/JourneySetupController.scala]: Do you mean on the cases? I always thought if it was more than one line function inside you had to have them. I'll take them off then.

**Comment by davidmthornton** [comment]
All changes made and also added a try catch for creation of the ContinueUrl because that was throwing exceptions as well.

Cheers.

File: app/controllers/JourneySetupController.scala [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+  private[controllers] def extractRedirectUrl(json: JsValue): Option[String] = {
+    try{
+      Some((json \ "redirectUrl").as[String])
+    }catch{
+      case jsEx:JsResultException =>
+        Logger.warn("[JourneySetupController][extractRedirectUrl] Couldn't find redirectUrl key in json provided", jsEx)
+        None
+
+      case _:Throwable =>
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you change this to the below please:
case e: Exception =>

File: app/controllers/JourneySetupController.scala [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            }
+          }else{
+            Logger.warn("[JourneySetupController][journeySetup] redirectUrl was empty or an invalid absolute url")
+            Future.successful(BadRequest)
+          }
+
+        }catch{
+          case iae : IllegalArgumentException =>
+            Logger.warn("[JourneySetupController][journeySetup] couldn't create ContinueUrl from url provided.", iae)
+            Future.successful(BadRequest)
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you add a catch all after this one please 
e.g.
case e: Exception =>
case e: Exception =>

File: app/controllers/JourneySetupController.scala [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  private[controllers] def extractRedirectUrl(json: JsValue): Option[String] = {
+    try{
+      Some((json \ "redirectUrl").as[String])
+    }catch{
+      case jsEx:JsResultException =>
+        Logger.warn("[JourneySetupController][extractRedirectUrl] Couldn't find redirectUrl key in json provided", jsEx)
+        None
+
+      case _:Throwable =>
+        Logger.warn("[JourneySetupController][extractRedirectUrl] Encountered an unknown error")
```
Comment (HarryGull) [inline app/controllers/JourneySetupController.scala]: Can you pass the exception into the logger overload too please
